qcodes/instrument_drivers/Keysight/N52xx.py

```diff
@@ -0,0 +1,422 @@
+    """
+    def __init__(self,
+                 name: str,
+                 instrument: 'PNABase',
+                 sweep_format: str,
+                 label: str,
+                 unit: str,
+                 memory: bool=False) -> None:
+        super().__init__(name,
+                         instrument=instrument,
+                         label=label,
+                         unit=unit,
+                         setpoint_names=('frequency',),
+                         setpoint_labels=('Frequency',),
+                         setpoint_units=('Hz',)
+                         )
+        self.sweep_format = format
+        self.memory = memory
+
+    def get_raw(self) -> Sequence[float]:
+        root_instr = self._instrument.root_instrument
+        # Check if we should run a new sweep
+        if root_instr.auto_sweep():
+            prev_mode = self.run_sweep()
+        # Ask for data, setting the format to the requested form
+        self._instrument.write(f'CALC:FORM {self.sweep_format}')
+        data = np.array(root_instr.visa_handle.query_binary_values('CALC:DATA? FDATA', datatype='f', is_big_endian=True))
+        # Restore previous state if it was changed
+        if root_instr.auto_sweep():
+            root_instr.sweep_mode(prev_mode)
+
+        return data
+
+    def run_sweep(self) -> str:
+        root_instr = self._instrument.root_instrument
+        # Store previous mode
+        prev_mode = root_instr.sweep_mode()
+        # Take instrument out of continuous mode, and send triggers equal to the number of averages
+        if root_instr.averages_enabled:
+            avg = root_instr.averages()
+            root_instr.write('SENS:AVER:CLE')
+            root_instr.write('SENS:SWE:GRO:COUN {0}'.format(avg))
+            root_instr.root_instrument.sweep_mode('GRO')
+        else:
+            root_instr.root_instrument.sweep_mode('SING')
+
+        # Once the sweep mode is in hold, we know we're done
+        while root_instr.sweep_mode() != 'HOLD':
+            time.sleep(0.1)
+
+        # Return previous mode, incase we want to restore this
+        return prev_mode
+
+class PNAPort(InstrumentChannel):
+    """
+    Allow operations on individual PNA ports.
+    Note: This can be expanded to include a large number of extra parameters...
+    """
+
+    def __init__(self,
+                 parent: 'PNABase',
+                 name: str,
+                 port: int,
+                 min_power: Union[int, float],
+                 max_power: Union[int, float]) -> None:
+        super().__init__(parent, name)
+
+        self.port = int(port)
+        if self.port < 1 or self.port > 4:
+            raise ValueError("Port must be between 1 and 4.")
+
+        pow_cmd = f"SOUR:POW{self.port}"
+        self.add_parameter("source_power",
+                           label="power",
+                           unit="dBm",
+                           get_cmd=pow_cmd + "?",
+                           set_cmd=pow_cmd + "{}",
```

Comment:
Should there be a whitespace between the  pow_cmd and the value when setting like in the commands below. 
I would personally find it clearer to write this as: `get_cmd=f'{pow_cmd}?'` and `set_cmd=f'{pow_cmd} {{}}'`

Reply:
Done